### ISSUE TITLE
Initial contextUpdate request payload has impact on subsequent contextUpdate requests in case they are shorter strings

### ISSUE THREAD

**Issue by manubaum** [telefonicaid/fiware-orion#1500]
Scenario:
- latest Orion broker freshly started (standard centOs rpm installation)
- mongoDB database empty
- two HTTP requests with the following payload are sent to /ngsi10/updateContext

Payload 1:
{"contextElements":[{"type":"TypeAddedFirst","id":"10","attributes":[{"name":"attr1","type":"int","value":"10"}]}],"updateAction":"APPEND"}

This request succeeds.

Payload 2:
{"contextElements":[{"type":"Second","id":"123","attributes":[{"name":"attr2","type":"int","value":"10"}]}],"updateAction":"APPEND"}

This request fails with a 'JSON Parse Error', because there is a "P" at the end of the payload 2, which originates from the first request (payload 1), since the character at index 132 from payload one is somehow considered also in payload 2. It is likely a buffer-index bug.

This is the output form the orion:
time=2015-11-13T11:54:33.881CET | lvl=DEBUG | trans=1447412040-197-00000000003 | function=jsonTreat | comp=Orion | msg=jsonRequest.cpp[191]: Treating a JSON request: '{"contextElements":[{"type":"Second","id":"123","attributes":[{"name":"attr2","type":"int","value":"10"}]}],"updateAction":"APPEND"}P'

**Comment by kzangeli** [comment]
Yeah, this is a new bug that we discovered and fixed.
In develop it is fixed already, see PR #1450

**Comment by fgalan** [comment]
Looking to @kzangeli 's cited PR, it seems that the fix was included in Orion 0.25.0 (released on November 2nd).

Could you tell us which Orion version are you using, please?

**Comment by manubaum** [comment]
Thanks for the fast replies. We use 0.25.0 (verified by both yum info and contextBroker --version). So either the PR #1450 did not fix this or it is not part of 0.25.0. Is there any workaround known for this issue since it is a blocker on our side (and I guess for many other users too, because it is hard to detect and kinda breaks seemingly random).

**Comment by kzangeli** [comment]
The fix was pretty simple, a one-liner ... However, it is supposed to be in the 0.25.0 release ...
Need to look into this. Asap ...
No 'simple work-around' possible, the source code needs to be modified.

**Comment by fgalan** [comment]
After a more careful examination, the bug was in 0.25.0 but it has been already fixed in develop (it correspond to this entry in the CHANGELOG_NEXT_RELEASE: https://github.com/telefonicaid/fiware-orion/blob/develop/CHANGES_NEXT_RELEASE#L9).

If you cannot wait until 0.26.0 gets released (by the end of November), I can provide an "interim" RPM with the fix if compiling from sources is too complex in your case.

**Comment by manubaum** [comment]
I try to build develop and if I fail I'd be happy to get the RPM.

**Comment by fgalan** [comment]
You can donwload from here:

http://www.filedropper.com/contextbroker-025020151112140227-1x8664

(Given it is not an official Orion version, I cannot use the offical FIWARE yum repository. Not sure if filedropper is the best way of sharing files... if you know a better alternative, please tell me).

**Comment by manubaum** [comment]
So we built from develop and we see the bug fixed.

**Comment by fgalan** [comment]
After last comment from @manubaum , I understand this issue can be closed.

**Comment by manubaum** [comment]
This problem was not fully fixed with the 0.26.0 (nor the 0.26.1). We experienced the same problem again when calling v1/subscribeContext?notifyFormat=JSON multiple times. Again, the problem is, that in our context subscription requests, characters are showing up we did not send to the orion context broker (note the 'a' character after the JSON object)

time=2016-01-06T17:35:07.467CET | lvl=DEBUG | trans=1452096388-828-00000000109 | function=jsonTreat | comp=Orion | msg=jsonRequest.cpp[191]: Treating a JSON request: '{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["event","data"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"EVENTNAME","isPattern":"true"}],"attributes":["event","data"]}a' 

So we switched back to the own compiled version:
0.25.0-next (git version: 6e99d1068b49feb8d0f98e14c5de0972069b7abd)

**Comment by fgalan** [comment]
Thank you for the feeback. We will reopened the issue while we debug on it.

In order to reproduce the bug, could you provide the exact payloads you are using (as you did in the updateContext case in the original bug description), please?

**Comment by fgalan** [comment]
We have tried to reproduce the problem, without success. Detail follows.

CB run using:

```
./contextBroker -fg -noCache  -logLevel DEBUG -t 0-255
```

Payloads:

```
$ cat payload_long.json 
{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["event","data"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"EVENTNAME","isPattern":"true"}],"attributes":["event","dataxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx”]}

$ cat payload_short.json 
{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["e","d"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"E","isPattern":"true"}],"attributes":["e","d"]}
```

Curl-based test: first the long payload, next the short payload. No "overlap" is observed in the last case and no "JSON Parse Error".

```
$ curl -i localhost:1026/v1/subscribeContext?notifyFormat=JSON -H "Content-type: application/json" -H "Accept: application/json" --data @payload_long.json
HTTP/1.1 200 OK
Content-Length: 108
Content-Type: application/json
Date: Thu, 21 Jan 2016 11:01:53 GMT

{
  "subscribeResponse" : {
    "subscriptionId" : "56a0baa1f56d3a25d93d7c7b",
    "duration" : "P1M"
  }
}

$ curl -i localhost:1026/v1/subscribeContext?notifyFormat=JSON -H "Content-type: application/json" -H "Accept: application/json" --data @payload_short.json
HTTP/1.1 200 OK
Content-Length: 108
Content-Type: application/json
Date: Thu, 21 Jan 2016 11:01:58 GMT

{
  "subscribeResponse" : {
    "subscriptionId" : "56a0baa6f56d3a25d93d7c7c",
    "duration" : "P1M"
  }
}
```

Logs traces (only the "Treating a JSON Request"):

```
DEBUG@11:11:21  RestService.cpp[413]: Parsing payload for URL '/v1/subscribeContext', method 'POST', service vector index: 69
DEBUG@11:11:21  RestService.cpp[415]: Parsing payload '{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["event","data"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"EVENTNAME","isPattern":"true"}],"attributes":["event","dataxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"]}'
DEBUG@11:11:21  RestService.cpp[91]: parsing data for service 'SubscribeContextRequest'. Method: 'POST'
DEBUG@11:11:21  RestService.cpp[92]: outFormat: JSON
DEBUG@11:11:21  jsonRequest.cpp[151]: Found jsonRequest of type 13, method 'POST' - index 8 (/entities)
DEBUG@11:11:21  jsonRequest.cpp[195]: Treating a JSON request: '{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["event","data"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"EVENTNAME","isPattern":"true"}],"attributes":["event","dataxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"]}'

[…]

DEBUG@11:13:48  RestService.cpp[413]: Parsing payload for URL '/v1/subscribeContext', method 'POST', service vector index: 69
DEBUG@11:13:48  RestService.cpp[415]: Parsing payload '{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["e","d"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"E","isPattern":"true"}],"attributes":["e","d"]}'
DEBUG@11:13:48  RestService.cpp[91]: parsing data for service 'SubscribeContextRequest'. Method: 'POST'
DEBUG@11:13:48  RestService.cpp[92]: outFormat: JSON
DEBUG@11:13:48  jsonRequest.cpp[151]: Found jsonRequest of type 13, method 'POST' - index 8 (/entities)
DEBUG@11:13:48  jsonRequest.cpp[195]: Treating a JSON request: '{"reference":"http://URL/contextUpdate.json","duration":"P1M","notifyConditions":[{"condValues":["e","d"],"type":"ONCHANGE"}],"entities":[{"id":".*","type":"E","isPattern":"true"}],"attributes":["e","d"]}'
```

Test has been done with the last version in develop at this moment (0.26.1-next 6dda36dd7415403a8678acb24dbd1efa77ea9fba).

In summary, we are not able to reproduce the problem. @manubaum could you tell us the exact way of reproducing it in your case (i.e. providing the same information we have provided above: CB run line, payloads, curl invocation, etc.)? Thanks!

**Comment by fgalan** [comment]
(Lowering priority to P4, given to un-reproductibility)
